### PR TITLE
[WIP] Reimplement mark all as read using watermarks

### PR DESCRIPTION
**Pull request by kytrinyx**
**Note: this is a follow-on pull request building on four PRs (#3141, #3142, and #3131 which are separate, and  #3143 which builds on #3131). All of these should be merged before this one. This will make it easier to review, as it will contain significantly fewer changes.**

The earlier PRs add a watermarks table to the database, and use both the watermark records and the view records to set the read/unread state on the streams.

Now that all that prior functionality is in place, we can start using the watermarks table for the "mark all as read" feature, rather than creating individual view records for every single exercise in the stream.

This should alleviate some of the pressure on our database, which is quickly approaching the 10 million row limit.

See https://github.com/exercism/exercism.io/issues/3130

### CODE REVIEW

File: lib/exercism/team_stream_filters.rb [hunk 1]
```diff
@@ -28,8 +36,14 @@ def views_by_id
...
 
-    def rows
-      execute(sql)
+    def watermarked_by_id
+      @watermarked_by_id ||= execute(watermarks_sql).each_with_object(Hash.new(0)) do |row, watermarked|
+        watermarked[row["id"]] = row["total"].to_i
+      end
+    end
+
+    def watermarks_sql
```
Comment (bernardoamc) [inline lib/exercism/team_stream_filters.rb]: Is this WIP?

File: lib/exercism/team_stream_filters.rb [hunk 2]
```diff
@@ -72,15 +86,35 @@ def sql
...
+          ON ex.id=views.exercise_id
+        WHERE ex.archived='f'
+          AND ex.iteration_count > 0
+          AND ex.user_id IN (#{user_ids_param})
+          AND views.user_id=#{viewer_id}
+          AND views.last_viewed_at > ex.last_activity_at
+      SQL
+    end
+
+    def watermarks_sql
```
Comment (bernardoamc) [inline lib/exercism/team_stream_filters.rb]: Oh, here is the real method. :D

File: lib/exercism/team_stream_filters.rb (hunk 1 above)
Comment (bernardoamc) [inline lib/exercism/team_stream_filters.rb]: Got it.

**Comment by bernardoamc** [comment]
Awesome PR! I was wondering if we could optimize the SQL and/or refactor the filters, going to take a stab at this at some point. 👍

File: lib/exercism/team_stream_filters.rb (hunk 1 above)
Comment (kytrinyx) [inline lib/exercism/team_stream_filters.rb]: Yeah, sorry. I should put a comment in there.

**Comment by kytrinyx** [comment]
I've been thinking about how to refactor and had some ideas that I wasn't convinced were good. I'd love some help with that whole piece! I think the streams are only going to grow. We have the team streams, the track streams, and soon (I hope) the profile streams, which should also include comments.

**Pull request merged**